transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -0,0 +1,267 @@
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                if (previousBuf != null) {
+                    ctx.write(previousBuf, ctx.voidPromise());
+                }
+                break;
+            }
+            try {
+                if (entry instanceof ByteBuf) {
+                    if (previousBuf != null) {
+                        ctx.write(previousBuf, ctx.voidPromise());
+                    }
+                    previousBuf = (ByteBuf) entry;
+                } else if (entry instanceof ChannelPromise) {
+                    ctx.write(previousBuf, (ChannelPromise) entry);
+                    previousBuf = null;
+                } else {
+                    ctx.write(previousBuf).addListener((ChannelFutureListener) entry);
+                    previousBuf = null;
+                }
+            } catch (Throwable t) {
+                if (pending == null) {
+                    pending = t;
+                }
+            }
+        }
+        if (pending != null) {
+            throw new IllegalStateException(pending);
+        }
+    }
+
+    /**
+     * Calculate the result of {@code current + next}.
+     */
+    protected abstract ByteBuf compose(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf next);
+
+    /**
+     * Calculate the first {@link ByteBuf} which will be used in subsequent calls to
+     * {@link #compose(ByteBufAllocator, ByteBuf, ByteBuf)}.
+     */
+    protected ByteBuf composeFirst(ByteBufAllocator allocator, ByteBuf first) {
+        return first;
+    }
+
+    /**
+     * The value to return when {@link #remove(ByteBufAllocator, int, ChannelPromise)} is called but the queue is empty.
+     * @return the {@link ByteBuf} which represents an empty queue.
+     */
+    protected abstract ByteBuf removeEmptyValue();
+
+    /**
+     * Get the number of elements in this queue added via one of the {@link #add(ByteBuf)} methods.
+     * @return the number of elements in this queue.
+     */
+    protected final int size() {
+        return bufAndListenerPairs.size();
+    }
+
+    private void releaseAndCompleteAll(ChannelFuture future) {
+        readableBytes = 0;
+        Throwable pending = null;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                break;
+            }
+            try {
+                if (entry instanceof ByteBuf) {
+                    ReferenceCountUtil.safeRelease(entry);
+                } else {
+                    ((ChannelFutureListener) entry).operationComplete(future);
+                }
+            } catch (Throwable t) {
+                pending = t;
+            }
+        }
+        if (pending != null) {
+            throw new IllegalStateException(pending);
+        }
+    }
+
+    private void checkByteBufPreconditions(ByteBuf buf) {
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
```

Comment:
`buf.readableBytes()` should not be `<0` but since flipping signs to detect overflow is part of the JLS [1] this should be safe and a bit more general. thanks for the suggestion.

https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.18.2
```
If an integer addition overflows, then the result is the low-order bits of the mathematical sum as represented in some sufficiently large two's-complement format. If overflow occurs, then the sign of the result is not the same as the sign of the mathematical sum of the two operand values.
```